src/py21cmmc/emulator.py

```diff
@@ -0,0 +1,196 @@
+        elif type(params) == np.ndarray:
+            theta = params.copy()
+        else:
+            raise TypeError('theta is in the wrong format. Should be AstroParams object or nine astrophysical parameters in same order as param_keys.')
+        if len(theta.shape) == 1:
+            theta = theta.reshape([1,-1])
+        normed = True
+        # Check that theta is normalized, if not, normalise it.
+        if isinstance(params, p21.AstroParams) or max(theta.ravel()) > 1 or min(theta.ravel()) < 0:
+            log.info('Theta is not normalized.')
+            normed = False # to indicate that input params was not normalised
+            for i in range(theta.shape[-1]):
+                if i == 0 or i == 2 or i == 4 or i == 6:
+                    theta[:,i] = np.log10(theta[:,i])
+                if i == 7:
+                    theta[:,i] /= 1000. #E0 in keV
+                theta[:,i] -= self.limits[i,0]
+                theta[:,i] /= (self.limits[i,1] - self.limits[i,0])
+        log.info('Normed params: %s' % theta)
+        log.info('Start emulator prediction...')
+        emu_pred = self.model.predict(theta, verbose = False)
+        log.info('End emulator prediction.')
+
+        Tb_pred_normed = emu_pred[:,:84]
+        xHI_pred = emu_pred[:,84:84*2]
+        Ts_pred_normed = emu_pred[:,2*84:84*3]
+        Ts_undefined_pred = emu_pred[:,84*3]
+        PS_pred_normed = emu_pred[:,84*3+1:].reshape((theta.shape[0], 60,12))
+
+        # Set the xHI < z(Ts undefined) to 0
+        xHI_pred_fix = np.zeros(xHI_pred.shape)
+
+        tau = np.zeros(theta.shape[0])
+        uvlfs = np.zeros((theta.shape[0], 3, len(self.uv_lf_zs),100))
+        for i in range(theta.shape[0]):
+            zbin = np.argmin(abs(self.zs - Ts_undefined_pred[i]))
+            if xHI_pred[i,zbin] < 1e-1:
+                xHI_pred_fix[i,zbin:] = xHI_pred[i,zbin:]
+            else:
+                xHI_pred_fix[i,:] = xHI_pred[i,:]
+            # Use py21cmFAST to analytically calculate UV LF and $\tau_e$
+            log.info('Begin tau computation...')
+            tau[i] = p21.wrapper.compute_tau(redshifts = self.zs, global_xHI = xHI_pred_fix[i,:])
+            log.info('Tau computation completed successfully.')
+
+            # Build astro_params array from theta \in [0,1]
+            if (isinstance(params, p21.AstroParams) or type(params) == dict) and not normed:
+                uvlfs[i,...] = np.array(p21.wrapper.compute_luminosity_function(redshifts = self.uv_lf_zs, astro_params=params))
+            else:
+                # Restore dimensions i.e. undo the limits
+                log.info('Restore dimensions to theta to calculate UV LF.')
+                theta_wdims = np.zeros(theta.shape)
+                for j in range(theta.shape[-1]):
+                    theta_wdims[:,j] = self.limits[j, 0] + theta[:, j] * (self.limits[j, 1] - self.limits[j, 0])
+                    if j == 7:
+                        theta_wdims[:,j] = 1000. * theta_wdims[:,j]
```

Comment:
also, I thought u are emulating tau and UVLF. But it turns out you are not. Then why go all the trouble modifying them in likelihood.py? Just do the calculation there

Reply:
Because Andrei wants them returned by the emulator (yes, even though they are not emulated). So since the calculation is already done inside, I don't want to do it twice.